Task/gcndsp 304 end user portal search input split

ChristopherJamesMorris opened the pull request:
This PR also combines the changes made by Lee for NoResults

ChristopherJamesMorris committed:
Splitting the logic to be able to search by Ods Code or Name from different screens

ChristopherJamesMorris committed:
Changed SearchModeEnums to SearchMode

ChristopherJamesMorris committed:
Combined changes for NoResults with work to split the code and name based search

ChristopherJamesMorris committed:
Final amendments for spltting the search by code and name

ChristopherJamesMorris committed:
Minor styling change for NoResults

leew commented on modules/end-user-portal/src/Core/HttpClientServices/SiteService.cs:
```diff
@@ -7,48 +7,61 @@ namespace GpConnect.NationalDataSharingPortal.EndUserPortal.Core.HttpClientServi
-        }
-        result = JsonConvert.DeserializeObject<List<SearchResultEntry>>(x.Result, _options);
-      });
+    private static ILogger<SiteService> _logger;
```
Should this be read only rather than static?

ChristopherJamesMorris replied:
I've made this read only now

leew commented on modules/end-user-portal/src/Core/HttpClientServices/SiteService.cs:
```diff
@@ -7,48 +7,56 @@ namespace GpConnect.NationalDataSharingPortal.EndUserPortal.Core.HttpClientServi
+            };
+            var result = default(List<SearchResultEntry>);
+
+            var url = UriExtensions.AddQueryParamsToObject("transparency-site", CreateQuery(searchRequest), _options);
```
@ChristopherJamesMorris sorry I think that we can use the simpler QueryHelper method now that we are only adding one param

public static string AddQueryString (string uri, string name, string value);

ChristopherJamesMorris replied:
Done.

leew commented on modules/end-user-portal/src/Core/QueryStringParameter.cs:
```diff
@@ -0,0 +1,11 @@
+namespace GpConnect.NationalDataSharingPortal.EndUserPortal.Core;
+
+public class QueryStringParameterAttribute : Attribute
```
Nice I really like this approach

ChristopherJamesMorris replied:
Thanks.

leew commented on modules/end-user-portal/src/Models/NoResultsModel.cs:
```diff
@@ -0,0 +1,26 @@
+    [BindProperty(SupportsGet = true)]
+    public SearchMode Mode { get; set; } = SearchMode.Name;
+
+    public string SearchType => Mode == SearchMode.Name ? SearchMode.Name.GetNameParameter() : SearchMode.Code.GetNameParameter();
```
this can call Mode.GetNameParameter()?

ChristopherJamesMorris replied:
Done.

leew commented on modules/end-user-portal/src/Pages/Search/Name.cshtml:
```diff
@@ -0,0 +1,53 @@
+@using Microsoft.AspNetCore.Mvc.ModelBinding
+@model GpConnect.NationalDataSharingPortal.EndUserPortal.Pages.Search.SearchByNameModel
+@{
+    var providerNameInvalid = Model.ModelState["ProviderName"]?.ValidationState == ModelValidationState.Invalid;
```
I don't understand this code, I would like to discuss this please

ChristopherJamesMorris replied:
On ProviderName in the Model, there is a Required data annotation. If the user clicks on Search without entering a value for ProviderName the modelstate would be invalid.

This code traps whether it is invalid or not and if it's the former, the UI then displays the error markup which you can see further down in the HTML

leew commented on modules/end-user-portal/src/Models/NoResultsModel.cs:
```diff
@@ -0,0 +1,26 @@
+
+    public string SearchType => Mode == SearchMode.Name ? SearchMode.Name.GetNameParameter() : SearchMode.Code.GetNameParameter();
+
+    [Display(Name = "OrganisationNameNoResults", ResourceType = typeof(DataFieldNameResources))]
```
Can we move this to the SearchMode enum please

ChristopherJamesMorris replied:
Sorted.

leew commented on modules/end-user-portal/src/Pages/Search/NoResults.cshtml:
```diff
@@ -0,0 +1,26 @@
+                    </h3>
+                    <p>There are no results for this @Model.SearchType</p>
+                    <p>You can search again by:
+                        <br/><a class="nhsuk-card__link" asp-page="/search/name" asp-route-ProviderName=@Model.NameQueryOrNull>@Html.DisplayNameFor(x => x.NameQueryOrNull)</a>
```
Can we pull this from the SearchMode enum please

ChristopherJamesMorris replied:
Sorted.

ChristopherJamesMorris committed:
Further use of the Search Mode enum to display labels on No Results
Minor changes to No Results page
Removed now defunct entries from Resources file

ChristopherJamesMorris committed:
Removed unnecessary code

leew commented on modules/end-user-portal/src/Core/HttpClientServices/SiteService.cs:
```diff
@@ -29,7 +30,7 @@ public async Task<List<SearchResultEntry>> SearchSitesAsync(SearchRequest search
             var result = default(List<SearchResultEntry>);
 
-            var url = UriExtensions.AddQueryParamsToObject("transparency-site", CreateQuery(searchRequest), _options);
+            var url = QueryHelpers.AddQueryString("transparency-site", searchRequest.Mode.GetQueryStringParameter(), searchRequest.Query);
```
Nice!
